subprojects/core-api/src/main/java/org/gradle/api/artifacts/transform/TransformOutputs.java

```diff
@@ -35,7 +35,31 @@
 public interface TransformOutputs {
     /**
      * Registers an output directory.
-     * For a relative path, a location for the output directory is provided.
+     *
+     * <p>
+     *     For an absolute path, the location is registered as an output directory of the {@link TransformAction}.
+     *     The path is required to point to the {@link InputArtifact} or be inside it if the input artifact is a directory.
+     *     Example:
+     * </p>
+     * <pre>
+     * Provider&lt;FileSystemLocation&gt; inputArtifact = ...
+     * outputs.dir(inputArtifact.get().getAsFile());
+     * outputs.dir(new File(inputArtifact.get().getAsFile(), "sub-dir"));
+     * </pre>
+     *
+     * <p>
+     *     For a relative path, a location for the output directory is provided by Gradle, so that the {@link TransformAction} can produce its outputs at that location.
+     *     The directory is created automatically when calling the method.
+     *     Example:
+     * </p>
+     * <pre>
+     * File myOutput = outputs.dir("my-output");
+     * Files.write(myOutput.toPath().resolve("file.txt"), "Generated text");
+     * </pre>
+     *
+     * <p>
+     *     The registered directory needs to exist when the {@link TransformAction#transform(TransformOutputs)} method finishes.
```

Comment:
I would turn this around and say:

```suggestion
     *     <strong>Note:</strong> it is an error if the registered directory does not exist when the {@link TransformAction#transform(TransformOutputs)} method finishes.
```